applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_sparse_utility.cpp

```diff
@@ -0,0 +1,129 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Manuel Messmer
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "custom_utilities/perturb_geometry_sparse_utility.h"
+#include "custom_utilities/node_search_utility.h"
+#include "utilities/builtin_timer.h"
+#include "utilities/parallel_utilities.h"
+
+namespace Kratos
+{
+
+int PerturbGeometrySparseUtility::CreateRandomFieldVectors(){
+    KRATOS_TRY;
+
+    int number_of_nodes = mrInitialModelPart.NumberOfNodes();
+
+    BuiltinTimer assemble_cl_matrix_timer;
+
+    // Construct and initialize required containers
+    ModelPart::NodesContainerType nodes = mrInitialModelPart.Nodes();
+    ResultNodesContainerType  results;
+    SparseMatrixType correlation_matrix;
+    correlation_matrix.resize(number_of_nodes,number_of_nodes,false);
+    TSparseSpaceType::SetToZero(correlation_matrix);
+
+    // Construct and initialize searcher
+    NodeSearchUtility searcher(nodes);
+
+    // Assemble correlation matrix
+    int counter = 0;
+    const auto it_node_begin = mrInitialModelPart.NodesBegin();
+
+    for( int i = 0; i < number_of_nodes; i++){
+        auto it_node = it_node_begin + i;
+        // Add current nodes to results
+        results = { mrInitialModelPart.Nodes().GetContainer()[i] };
+        // Find all neighbouring nodes
+        searcher.SearchNodesInRadius(*it_node, 3.0*mCorrelationLength, results);
+        for( size_t j = 0; j < results.size(); j++){
+            counter++;
+            int index = results[j]->GetId() - 1;
+            auto it_node_results = it_node_begin + index;
+
+            correlation_matrix(i, index) =  CorrelationFunction( *it_node, *it_node_results, mCorrelationLength);
+        }
```

Comment:
this can also easily be parallalized with #7261 

(actually I added an example for this usecase)

Reply:
The problem here is that the inner for loop is of varying size.

Reply:
i meant the outter loop over the nodes